### ISSUE TITLE
pod stuck in terminating and volume can't be reclaimed

### ISSUE THREAD

**Issue by gaorong** [kubernetes/kubernetes#71400]
**What happened**:
When our k8s slave node power down accidently,  we reboot it and kubelet restarts. then there are some error logs as below: 
```
I1125 01:06:20.270821   28282 reconciler.go:154] Reconciler: start to sync state
W1125 01:06:20.278163   28282 reconciler.go:394] Could not construct volume information, cleanup the mounts. (pod.UID fa90db70-f00a-11e8-8693-000c292d53d3, volume.SpecName lvm): no volume plugin matched
E1125 01:06:20.278214   28282 reconciler.go:442] volumeHandler.UnmountVolumeHandler for UnmountVolume failed for volume "" (UniqueName: "lvm") pod "fa90db70-f00a-11e8-8693-000c292d53d3" (UID: "fa90db70-f00a-11e8-8693-000c292d53d3") : UnmountVolume.FindPluginByName failed for volume "" (UniqueName: "lvm") pod "fa90db70-f00a-11e8-8693-000c292d53d3" (UID: "fa90db70-f00a-11e8-8693-000c292d53d3") : no volume plugin matched
```
it turns out kubelet invoke `FindPluginByName` failed.  

In production environment, we have a flexvolume driver located in `/usr/libexec/kubernetes/kubelet-plugins/volume/exec/xiaomi~lvm/lvm`. kubelet loads this driver and names as `flexvolume-xiaomi/lvm`(see [here](https://github.com/kubernetes/kubernetes/blob/7098f1ad383b84bc8e807b049ab42629a70bb04e/pkg/volume/flexvolume/plugin.go#L117-L119)). when we create a pod, kubelet will mount volume at directory such as `/var/lib/kubelet/pods/fa90db70-f00a-11e8-8693-000c292d53d3/volumes/xiaomi~lvm/lvm`. After we reboot the machine, the kubelet trys to reconstruct volume information by scaning the volume directories. In this case, the volume plugin name scaned from directory is `xiaomi/lvm`(see [here](https://github.com/kubernetes/kubernetes/blob/7098f1ad383b84bc8e807b049ab42629a70bb04e/pkg/kubelet/volumemanager/reconciler/reconciler.go#L670-L688)). Obeviously, this name mismatched the name detected from `--volume-plugin-dir` previously, so the `FindPluginByName` and `UnmountVolume` function failed, the volume directory can't be deleted and the pod will stay terminating state.

**What you expected to happen**:
pods get terminated, volume resource can be reclaimed   

**How to reproduce it (as minimally and precisely as possible)**:
```
# create a pod with flexvolume
[root@k8s-test-master01 test]# cat lvm-test.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: nginx-lvm-demo
  namespace: default
spec:
  containers:
  - name: lvm
    image: nginx:latest
    volumeMounts:
    - name: lvm
      mountPath: /data
    ports:
    - containerPort: 80
  volumes:
  - name: lvm
    flexVolume:
      driver: xiaomi/lvm
      options:
        DiskSize: 20M

[root@k8s-test-master01 test]# kubectl create -f lvm-test.yaml 
pod "nginx-lvm-demo" created
[root@k8s-test-master01 test]# kubectl get pods
NAME              READY     STATUS    RESTARTS   AGE
nginx-lvm-demo   1/1       Running   0          8s

# stop kubelet and docker to emulat the machine down, must stop both docker and kublet
[root@k8s-test-master01 test]# systemctl stop kubelet docker
[root@k8s-test-master01 test]# kubectl get node
NAME      STATUS     ROLES     AGE       VERSION
node1     NotReady   <none>    17h       v1.14.0-alpha.0.560+94759c165f2320

# delete pod when the machine is still down
[root@k8s-test-master01 test]# kubectl delete pods nginx-lvm-demo
pod "nginx-lvm-demo" deleted
[root@k8s-test-master01 test]# kubectl get pods
NAME              READY     STATUS        RESTARTS   AGE
nginx-lvm-demo   1/1       Terminating   0          1m

# reboot the machine and pod still stuck in terminating
[root@k8s-test-master01 test]# systemctl restart kubelet docker
[root@k8s-test-master01 test]# kubectl get pods
NAME              READY     STATUS        RESTARTS   AGE
nginx-lvm-demo   0/1       Terminating   0          2m
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): both v1.10.2 and v1.14.0-alpha.0.560+94759c165f2320
- OS (e.g. from /etc/os-release): centos7
- Kernel (e.g. `uname -a`): Linux k8s-test-master01 3.10.0-862.el7.x86_64 #1 SMP Fri Apr 20 16:44:24 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: 
- Others:

/kind bug

**Comment by xiegang1** [comment]
This bug caused a lot of lvm leaks on our online system and should be a critical one. Can anyone could help have a review about this issue?

**Comment by GaoRong** [comment]
PTAL @jingxu97

**Comment by jingxu97** [comment]
Hi @Gaogong, I will help to fix this issue. The pod volume mount path is set up differently for flexvolume with other plugins.

**Comment by GaoRong** [comment]
@jingxu97  I agree with you, Could you explain more specifically where is the compatibility issue?  

AFAIK, there are many companies using flexvolume in production and many issues related terminating pod, If the compatibility is not very critical, maybe we should fix this issue first?

**Comment by jingxu97** [comment]
Currently kubelet volume manager create a pod directory and mount the volume there. For most of plugins, the directory name is a combination of pod id, plugin name and volume name. So when kubelet restarts, volume reconstruction code will try to scan the directory and retrieve the pod id, plugin name and volume name from the directory name. However, flexvolume is a little bit different. The flexvolume plugin name has a prefix of the driver name "flexvolume-xxx". So kubelet operator cannot match the correctly plugin name during volume reconstruction. 

If we change flex volume path to use the full plugin name instead of only the driver name, during cluster upgrade, the new version of kubelet cannot correctly get the path of the existing pods because it will use the new path and fail to clean up.

**Comment by gaorong** [comment]
Maybe I didn't explain it clearly, I'm sorry for my poor English.

There is no compatibility problem in my opinion. Kubelet doesn't know the flexvolume path now. this is why the error "can't find plugin" is always reported after reconstruction, we can reproduce this issue easily. After this fix, everything will be ok.

**Comment by jingxu97** [comment]
@gaorong maybe it is easier if we can talk directly. You can email me your contact information jinxu at google.com
Thanks!

**Comment by chakri-nelluri** [comment]
@jingxu97 this seems a bug and we should ideally fix it. Initial flexvolumes did not use this logic and it was later added as part https://github.com/kubernetes/kubernetes/pull/46382.

**Comment by chakri-nelluri** [comment]
Also @gaorong inline flexvolumes are not recommended and should be avoided. The volume name in inline flexvolumes is not unique and it could potentially conflict with other volumes in other pod specs.

**Comment by chakri-nelluri** [comment]
@verult, @jingxu97  it is better to remove the "flexvolume-" attribute rather than confuse the reconciler. This was supposed to be a logging fix, but seems it broke the driver on Kubelet restart. Might need a quicker fix and may be patch back the previous releases too dating back to 1.8.

**Comment by verult** [comment]
I agree with @jingxu97 that Flexvolume directory path format shouldn't be changed in order to avoid breaking cluster upgrades. I also agree with @chakri-nelluri that a logging fix shouldn't be favored if it causes breakage.

Here's an alternative (although maybe a little hacky): we know all volume plugins are either in-tree, Flex, or CSI. In-tree volumes are well-known, and CSI drivers are loaded differently. Thus reconstruction logic can keep a list of in-tree driver names + the csi in-tree plugin, and assume plugins with unrecognized names are Flex.

**Comment by verult** [comment]
Yeah it might be a fair amount of work. I'm OK with removing the "flexvolume-" prefix as long as we broadcast it in release notes

**Comment by chakri-nelluri** [comment]
Thanks @verult.
/assign verult

**Comment by gaorong** [comment]
@verult @chakri-nelluri  Sorry for not updating in time, this issue is being fixed in #71402, could you please help review?
